contracts/profit/src/state/mod.rs

```diff
@@ -1 +1,231 @@
+
+type IcaConnector = dex::IcaConnector<OpenIca, DexResponse<Idle>>;
+
+trait UpdateConfig
+where
+    Self: Sized,
+{
+    fn update_config(self, cadence_hours: u16) -> Self;
+}
+
+pub(crate) trait ProfitMessageHandler
+where
+    Self: Handler,
+{
+    fn confirm_open(
+        self,
+        deps: Deps<'_>,
+        env: Env,
+        _channel: Ics20Channel,
+        counterparty_version: String,
+    ) -> ContinueResult<Self> {
+        self.on_open_ica(counterparty_version, deps, env)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[repr(transparent)]
+#[serde(transparent)]
+pub(crate) struct State(StateEnum);
+
+#[derive(Serialize, Deserialize)]
+enum StateEnum {
+    OpenTransferChannel(OpenTransferChannel),
+    OpenIca(IcaConnector),
+    Idle(Idle),
+    BuyBack(StateLocalOut<BuyBack>),
+}
+
+impl State {
+    pub fn new(
+        querier: &QuerierWrapper<'_>,
+        config: Config,
+        connection_id: String,
+        oracle_addr: Addr,
+        time_alarms_addr: Addr,
+    ) -> ContractResult<Self> {
+        Ok(Self(StateEnum::OpenTransferChannel(
+            OpenTransferChannel::new(
+                config,
+                connection_id,
+                OracleRef::try_from(oracle_addr, querier)?,
+                TimeAlarmsRef::new(time_alarms_addr, querier)?,
+            ),
+        )))
+    }
+
+    pub fn load(storage: &dyn Storage) -> ContractResult<Self> {
+        STATE.load(storage).map_err(Into::into)
+    }
+
+    pub fn store(&self, storage: &mut dyn Storage) -> ContractResult<()> {
+        STATE.save(storage, self).map_err(Into::into)
+    }
+
+    pub fn try_update_config(self, cadence_hours: u16) -> ContractResult<Self> {
+        match self.0 {
+            StateEnum::OpenTransferChannel(transfer) => Ok(Self(StateEnum::OpenTransferChannel(
+                transfer.update_config(cadence_hours),
+            ))),
+            StateEnum::OpenIca(_) => Err(ContractError::unsupported_operation(
+                "Configuration changes are not allowed during ICA opening process.",
+            )),
+            StateEnum::Idle(idle) => Ok(Self(StateEnum::Idle(idle.update_config(cadence_hours)))),
+            StateEnum::BuyBack(_) => Err(ContractError::unsupported_operation(
+                "Configuration changes are not allowed during buy-back.",
+            )),
+        }
+    }
+
+    pub fn config(&self) -> ContractResult<&Config> {
+        match &self.0 {
+            StateEnum::OpenTransferChannel(transfer) => Ok(transfer.config()),
+            StateEnum::OpenIca(_) => Err(ContractError::unsupported_operation(
+                "Querying configuration is not allowed during ICA opening process.",
+            )),
+            StateEnum::Idle(idle) => Ok(idle.config()),
+            StateEnum::BuyBack(_) => Err(ContractError::unsupported_operation(
+                "Querying configuration is not allowed during buy-back.",
+            )),
+        }
+    }
+}
```

Comment:
Avoid using  `match` expressions to implement polymorphic behavior.  Use trait(s) implemented for all states, for example, `UpdateConfig`, and call its methods. The particular state would decide how to handle such requests.